src/crypter.rs

```diff
@@ -0,0 +1,145 @@
+use aes_gcm::{
+    aead::{AeadCore, AeadInPlace, KeyInit, OsRng},
+    Aes128Gcm as AesGcm, Error, KeySizeUser, Nonce,
+};
+use base64::{engine::general_purpose::STANDARD, DecodeError, Engine};
+use std::{
+    collections::VecDeque,
+    fmt::{self, Debug, Formatter},
+    iter,
+    str::FromStr,
+    sync::Arc,
+};
+use typenum::marker_traits::Unsigned;
+
+pub type Key = aes_gcm::Key<AesGcm>;
+
+#[derive(Clone)]
+pub struct Crypter(Arc<CrypterInner>);
+
+#[derive(thiserror::Error, Debug, Clone)]
+pub enum CrypterParseError {
+    #[error(transparent)]
+    Base64(#[from] DecodeError),
+
+    #[error("incorrect key length, must be {} bytes after base64 decode", <AesGcm as KeySizeUser>::KeySize::USIZE)]
+    IncorrectLength,
+
+    #[error("at least one key needed")]
+    Missing,
+}
+
+impl FromStr for Crypter {
```

Comment:
Rather than creating a Crypter from a string, I suggest we create Keys from a string, and Crypters from keys.

Reply:
What is your justification for this? Why create another type?

Reply:
To be clear, I'm suggesting `Key` implement `FromStr` (or otherwise we have a function that does this on a key-by-key basis). A `Crypter` is logically distinct from the keys it contains. This is a minor point, tbh.

Reply:
Currently, any type that is on `Config` needs to be `FromStr`, so we'd have to reconstruct a `Crypter` from `Keys` every time it was needed instead of storing a `Crypter` directly on `Config`. And the only purpose of `Keys` would be to construct a Crypter, as there is no need for any interface other than `From<&Keys> for Crypter`. This change would increase code without adding functionality.